data-products/src/main/scala/org/sunbird/obsrv/dataproducts/job/MasterDataProcessorIndexer.scala

```diff
@@ -0,0 +1,156 @@
+        e.printStackTrace()
+    }
+  }
+
+  def getPaths(datasource: DataSource): Paths = {
+    val dt = new DateTime(DateTimeZone.UTC).withTimeAtStartOfDay()
+    val timestamp = dt.getMillis
+    val date = dayPeriodFormat.print(dt)
+    val provider = providerFormat(config.getString("cloudStorage.provider"))
+    val cloudPrefix = if (config.getString("cloudStorage.provider") == "azure") {
+      provider.sparkProviderURIFormat + s"""://${config.getString("cloudStorage.accountName")}.blob.core.windows.net/${config.getString("cloudStorage.container")}/"""
+    } else {
+      provider.sparkProviderURIFormat + s"""://${config.getString("cloudStorage.container")}/"""
+    }
+    val pathSuffix = s"""masterdata-indexer/${datasource.datasetId}/$date/"""
+    val ingestionPath = cloudPrefix.replace(provider.sparkProviderURIFormat, provider.druidProviderPrefix) + pathSuffix
+    val datasourceRef = datasource.datasource + '-' + date
+    val outputFilePath = cloudPrefix + pathSuffix
+    Paths(datasourceRef, ingestionPath, outputFilePath, timestamp)
+  }
+
+  private def updateIngestionSpec(datasource: DataSource, datasourceRef: String, filePath: String): String = {
+    val deltaIngestionSpec = s"""{"type":"index_parallel","spec":{"dataSchema":{"dataSource":"$datasourceRef"},"ioConfig":{"type":"index_parallel"},"tuningConfig":{"type":"index_parallel","maxRowsInMemory":500000,"forceExtendableShardSpecs":false,"logParseExceptions":true}}}"""
+    val provider = providerFormat(config.getString("cloudStorage.provider"))
+    val inputSourceSpec = if(provider.druidProvider == "local")
+      s"""{"spec":{"ioConfig":{"type":"index_parallel","inputSource":{"type":"${provider.druidProvider}","baseDir":"${config.getString("cloudStorage.container")}","filter":"**.json.gz"}}}}"""
+    else s"""{"spec":{"ioConfig":{"type":"index_parallel","inputSource":{"type":"${provider.druidProvider}","objectGlob":"**.json.gz","prefixes":["$filePath"]}}}}"""
+    val deltaJson = parse(deltaIngestionSpec)
+    val inputSourceJson = parse(inputSourceSpec)
+    val ingestionSpec = parse(datasource.ingestionSpec)
+    val modIngestionSpec = ingestionSpec merge deltaJson merge inputSourceJson
+    compact(render(modIngestionSpec))
+  }
+
+  private def providerFormat(cloudProvider: String): BlobProvider = {
+    cloudProvider match {
+      case "local" => BlobProvider("file", "local", "file")
+      case "aws" => BlobProvider("s3a","s3", "s3")
+      case "azure" => BlobProvider("wasbs", "azure", "azure")
+      case "gcloud" => BlobProvider("gs", "google", "gs")
+      case "cephs3" => BlobProvider("s3a", "s3", "s3") // TODO: Have to check Druid compatibility
+      case "oci" => BlobProvider("s3a", "s3", "s3") // TODO: Have to check Druid compatibility
+      case _ => throw new Exception("Unsupported provider")
+    }
+  }
+
+  private def submitIngestionTask(ingestionSpec: String) = {
+    // TODO: Handle success and failure responses properly
```

Comment:
Why TODO? what else needs to be handled?

Reply:
Exceptions are handled. I will remove the commnet